### PR TITLE
Run doc_gram_verify target in lint CI job.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** infrastructure.

~~@jfehrle As is, the linter only checks that `fullGrammar` is up-to-date. In #14178, we recommend running the full `doc_gram_rsts` target. Should we extend `doc_gram_verify` to some sort of `doc_gram_rsts_verify`?~~ Done

Include changes originally in: #14178
Now on top of: #14504 (merged)

### CODE REVIEW

File: dev/lint-repository.sh
Comment (SkySkimmer) [inline dev/lint-repository.sh:40]:
```suggestion
(./configure -profile devel && make -j "$NJOBS" -f Makefile.make doc_gram_verify) || CODE=1
```

**Comment by jfehrle** [comment]
I don't see the need for 2 verify targets if that's what you mean.  I'll have to look at the code to make sure it still does what we want.  Also, for `make doc_gram_rsts`, the dependencies are incomplete because they're a bit circular, e.g. if you only change `insertprodn` in an rst, `make doc_gram_rsts` won't update the doc.  I usually touch `orderedGrammar` to get around this.

**Comment by Zimmi48** [comment]
> I don't see the need for 2 verify targets if that's what you mean.

That's not what I meant. What I meant was: what's currently implemented in this PR only tests that `fullGrammar` is up-to-date. Do we want to stop there or do we want to also test that `orderedGrammar` and the rst files are up-to-date?

I don't care if it's a new target that does this or if this is the `doc_gram_verify` one which is modified.

**Comment by jfehrle** [comment]
It should cover both fullGrammar/orderedGrammar and the rsts.

**Comment by Zimmi48** [comment]
Ah OK, that's not what the doc says: https://github.com/coq/coq/tree/master/doc/tools/docgram#how-to-use-the-tool

**Comment by Zimmi48** [comment]
I've just tested and I confirm that the doc is correct.

**Comment by jfehrle** [comment]
Where is the Dune/make rule for this?  I couldn't find it, I'd like to check the parameters it passes.

**Comment by Zimmi48** [comment]
It is in `Makefile.doc`. You were the one that wrote it BTW, but that was a long time ago :wink:

**Comment by jfehrle** [comment]
Multitasking makes you stupid. :wink:  I've been immersed in the debugger for the last week trying to get it ready for review and 8.14.

**Comment by jfehrle** [comment]
Ah, `make -f Makefile.dune world` doesn't compile `doc_grammar`.  Nor does `make doc_gram_rsts` or `make doc_gram_verify`.  I had to delete `doc_grammar.exe` from the `_build` directory.

**Comment by jfehrle** [comment]
The warnings always print because my code won't compile if I change it to:

```
let warn s =
  if !show_warn then begin
    Printf.eprintf "Warning: ";
    Printf.eprintf s
  end
```

The error shows up when `warn` is called, e.g.

```
          warn "%s: Duplicate production '%s: %s'\n" file nt (prod_to_str prod);

File "doc/tools/docgram/doc_grammar.ml", line 684, characters 10-14:
Error: This function has type (unit, out_channel, unit) format -> unit
       It is applied to too many arguments; maybe you forgot a `;'.
```

Any ideas on how to fix this cleanly?  The alternative is changing all the calls to `warn` like this:

```
        if !show_warn then
          warn "%s: Duplicate production '%s: %s'\n" file nt (prod_to_str prod);
```

```
$ ocaml --version
The OCaml toplevel, version 4.07.1
```

I never bothered to file a bug report.

**Comment by jfehrle** [comment]
> Should we extend `doc_gram_verify` to some sort of `doc_gram_rsts_verify`?

Just remove the `-short` from the command line.  I'll leave it to you to tweak `README.md`.

**Comment by Zimmi48** [comment]
The linter has been a sensitive topic in the past. So I'm wondering if it's the right place to put this call, or if it should rather be in one of the documentation jobs. cc @gares

**Comment by Zimmi48** [comment]
@SkySkimmer It looks like we're lacking a CODEOWNER for `dev/lint-repository.sh`. Should it be @coq/ci-maintainers?

**Comment by SkySkimmer** [comment]
That sounds reasonable.

**Comment by Zimmi48** [comment]
I'll do that in a separate PR.

**Comment by gares** [comment]
@Zimmi48 can you explain what this check does and how it is supposed to impact in our workflow? How do I detect a failure locally? How do I fix it?

**Comment by Zimmi48** [comment]
@gares This PR goes hand in hand with #14178 which adds a new PR template checkbox asking that if you have modified the grammar, then you run `make doc_gram_rsts`. If you do so and the target succeeds, then all you need to do is to commit the auto-generated changes and possibly also adapt the text of the documentation as you would anyway usually do. If it fails then it will generally mean that a change is needed in `common.edit_mlg`. Documentation about this file can be found in https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md, but it's also fine to just call @jfehrle or I for help.

**Comment by gares** [comment]
Makes sense to me.

**Comment by Zimmi48** [comment]
@jfehrle The linter has failed with `Error: doc/tools/docgram/fullGrammar is not current`. I am surprised because it works for me locally and it also worked in previous versions of the PR (before #14262). Do you have any idea what could be going on?

File: doc/tools/docgram/README.md
Comment (jfehrle) [inline doc/tools/docgram/README.md:115]: I find it strange to put the variable assignment at the beginning.  It seems to work, oddly enough, but I have no idea when the shell started to support that.

**Comment by jfehrle** [comment]
> Do you have any idea what could be going on?

Not really.  Two thoughts, though:
- If the sort in the wildcard expansion for DOC_MLGS is different from our local systems in CI, then `fullGrammar` won't match.  (One of first problems I ran into building Coq was that `awk` didn't work the same way in all environments, causing a compilation failure.)  Maybe print the expanded arguments and compare them to what you get locally.
- Remove the last line below in `doc_grammar.ml`, then print/diff the two files to look for clues.  The temp file name is just `fullGrammar.new` in the same directory as `fullGrammar`.

```
  let temp_file = (old_file ^ ".new") in
  if !exit_code <> 0 then
    Sys.remove temp_file
  else if args.verify then begin
    if not (files_eq old_file temp_file) then
      error "%s is not current\n" old_file;
    Sys.remove temp_file

**Comment by jfehrle** [comment]
> to just call @jfehrle or I for help.

should be "or me"

--Mr. Grammar Person

File: doc/tools/docgram/README.md
Comment (Zimmi48) [reply doc/tools/docgram/README.md:115]: I also tried at the end of the line, and oddly it didn't work for me. That's why I documented on the beginning of the line.

**Comment by jfehrle** [comment]
The humorist Dave Berry used to write "Mr. Language Person" columns from time to time. For example: 
https://webcache.googleusercontent.com/search?q=cache:9_hu14EvTxUJ:https://www.orlandosentinel.com/news/os-xpm-1997-04-29-9704260537-story.html+&cd=5&hl=en&ct=clnk&gl=us

**Comment by jfehrle** [comment]
I merged the dependency.

File: doc/tools/docgram/doc_grammar.ml
Comment (SkySkimmer) [inline doc/tools/docgram/doc_grammar.ml:1566]:
```suggestion
    if not (files_eq old_file temp_file) then begin
      error "%s is not current\n" old_file;
      ignore (Sys.command (sprintf "diff %s %s.new" old_file old_file))
    end;
```

File: Makefile.doc
Comment (SkySkimmer) [inline Makefile.doc:312]: Please add a comment about why we do this explicit listing

**Comment by Zimmi48** [comment]
Updated. @SkySkimmer do you want to be the assignee?

File: dev/lint-repository.sh
Comment (SkySkimmer) [inline dev/lint-repository.sh:40]: I'm not very happy with configuring here, is there a way we could use pure dune instead?

Comment (Zimmi48) [reply dev/lint-repository.sh:40]: Not as of today, this `_verify` target is more complete than the `@check-gram` one from Dune because the latter only looks at `fullGrammar` and `orderedGrammar`, not the `rst` files.

My experience with using Dune for the documentation has been very frustrating. IMHO in the future where we have a single build system, we should remove all the Dune files for documentation and just have a small Makefile for this package (which wouldn't require calling `configure` first because it would just call Dune to build the full Coq package before).

**Comment by Zimmi48** [comment]
@gares We had a segfault in the `build:base+async` job here. And also (that's more unusual) in a test from `test-suite:base`!

File: doc/tools/docgram/README.md
Comment (SkySkimmer) [reply doc/tools/docgram/README.md:115]: Variable assignments done as make arguments may not be mutated without `override` https://www.gnu.org/software/make/manual/make.html#Override-Directive
`foo=bar cmd` is just an environment variable so does not need override.

IMO the logic at https://github.com/coq/coq/blob/b0e85d75f8085b7e94d976a8243b7efe48bf9d0d/Makefile.doc#L51-L56 should set a new variable instead of mutating.

File: doc/tools/docgram/doc_grammar.ml
Comment (SkySkimmer) [inline doc/tools/docgram/doc_grammar.ml:1565]: Please use CUnix.sys_command instead

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (jfehrle) [inline .github/PULL_REQUEST_TEMPLATE.md:19]: There are more details, of course, but we shouldn't go overboard here.  For example, with the current doc_grammar, if someone adds a new command or tactic, the submitter will need to add it to the documentation.  We can't enable a lint check that every tactic appears in the doc until we've updated the syntax in the rest of the doc.  (Or similarly, a check that every tactic in the doc is in the code.)

File: CONTRIBUTING.md
Comment (jfehrle) [inline CONTRIBUTING.md:577]: Maybe mention how to handle added/removed commands and tactics here.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (Zimmi48) [reply .github/PULL_REQUEST_TEMPLATE.md:19]: Right, but reviewers have already been checking for this manually for many years now. So IMHO, this is just implicit in the "Added / updated documentation" general checkbox.

**Comment by Zimmi48** [comment]
I will implement the suggested change at https://github.com/coq/coq/pull/14245#discussion_r666859051 and https://github.com/coq/coq/pull/14245#discussion_r667120657.

**Comment by Zimmi48** [comment]
Updated to account for the two aforementioned comments.

Also now includes a commit updating `fullGrammar` and `common.edit_mlg` following #14562 (cc @herbelin). This doesn't change `orderedGrammar` and thus the documented syntax, but maybe it should?

**Review by jfehrle — APPROVED** [review]
Looks good to me.  (What could possibly go wrong?)  If you send an announcement to the developers when this is merged, feel free to mention I'm available to answer questions.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Comment by jfehrle** [comment]
@Zimmi48 Thanks for doing this!

**Pull request merged**
